### PR TITLE
Fix singleton-range `__setitem__` edge case

### LINKED ISSUE #12073: [BUG] Missing data with length-one sliced-based `__setitem__` and singleton array
**Describe the bug**

Trying to set a length-one slice of a series with a length-one array (rather than a scalar) loses the value.

**Steps/Code to reproduce bug**

```python
import cudf
import cupy
s = cudf.Series([1, 2, 3], dtype=cupy.int64)
s.iloc[:1] = cupy.asarray([7], dtype=cupy.int64)
print(s)
# 0    1 <------ expecting this to be 7
# 1    2
# 2    3
# dtype: int64
```

**Expected behavior**

Setting a length-one range with a length-one array should work.

***Why does this happen?***

There's an off-by-one in copying.copy_range:

```python
    if abs(target_end - target_begin) <= 1: # should be `<`.
        return target_column
```

### PR DESCRIPTION
**Pull request by wence-**
## Description

When trying to set a length-one range with a length-one array, an off-by-one error in `copying.copy_range` meant that the value was discarded. Fix that, and tidy up the semantics of `copy_range` a little while we're here. Closes #12073.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/copying.pyx [hunk 1]
```diff
@@ -132,31 +132,32 @@ def _copy_range(Column input_column,
...
+        source_end - source_begin == target_end - target_begin,
+        "Source and target ranges must be same length"
+    )
+    if target_end >= target_begin and inplace:
+        # FIXME: Are we allowed to do this when inplace=False?
         return target_column
 
-    if inplace is True:
+    if inplace:
         _copy_range_in_place(input_column, target_column,
```
Comment (wence-) [inline python/cudf/cudf/_lib/copying.pyx:161]: screwed up the renaming, sorry.

File: python/cudf/cudf/tests/test_setitem.py [hunk 2]
```diff
@@ -108,6 +108,12 @@ def test_series_set_item(psr, arg):
     assert_eq(psr, gsr)
 
 
+def test_series_setitem_singleton_range():
+    sr = cudf.Series([1, 2, 3], dtype=np.int64)
+    sr.iloc[:1] = np.asarray([7], dtype=np.int64)
+    assert_eq(sr, cudf.Series([7, 2, 3], dtype=np.int64))
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_setitem.py:114]:
```suggestion
    sr = cudf.Series([1, 2, 3], dtype=np.int64)
    psr = sr.to_pandas()
    np_arr = np.asarray([7], dtype=np.int64)
    sr.iloc[:1] = np_arr
    psr.iloc[:1] = np_arr
    assert_eq(sr, psr)
```

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_setitem.py:114]: Better to have a pandas comparison in the test?

Comment (wence-) [reply python/cudf/cudf/tests/test_setitem.py:114]: That just asserts we are bug-for-bug compatible with pandas though.

File: python/cudf/cudf/_lib/copying.pyx [hunk 3]
```diff
@@ -132,36 +132,37 @@ def _copy_range(Column input_column,
...
+    example via ``slice.indices``.
+    """
 
-    if target_begin > target_end:
+    assert (
+        source_end - source_begin == target_end - target_begin,
+        "Source and target ranges must be same length"
+    )
+    if target_end >= target_begin and inplace:
+        # FIXME: Are we allowed to do this when inplace=False?
```
Comment (wence-) [inline python/cudf/cudf/_lib/copying.pyx:157]: - [ ]TODO: address this comment.

Comment (wence-) [reply python/cudf/cudf/_lib/copying.pyx:157]: I'm happy to do so by just opening an issue to track in service of trying to get this in.

File: python/cudf/cudf/tests/test_setitem.py [hunk 4]
```diff
@@ -108,6 +108,16 @@ def test_series_set_item(psr, arg):
     assert_eq(psr, gsr)
 
 
+def test_series_setitem_singleton_range():
+    sr = cudf.Series([1, 2, 3], dtype=np.int64)
+    psr = sr.to_pandas()
+    value = np.asarray([7], dtype=np.int64)
+    sr.iloc[:1] = value
+    psr.iloc[:1] = value
+    assert_eq(sr, cudf.Series([7, 2, 3], dtype=np.int64))
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_setitem.py:117]:
```suggestion
```
This is now redundant, right?

Comment (wence-) [reply python/cudf/cudf/tests/test_setitem.py:117]: Well, only if we believe pandas is bug-free (I suppose it probably is here).

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_setitem.py:117]: In that case we could check for dtype explicitly with `check_dtype=True`:
```python
assert_eq(sr, psr, check_dtype=True)
```

Comment (wence-) [reply python/cudf/cudf/tests/test_setitem.py:117]: Done.

File: python/cudf/cudf/_lib/copying.pyx (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/_lib/copying.pyx:157]: @wence- could you open a separate issue to fix this todo?

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

File: python/cudf/cudf/_lib/copying.pyx (hunk 3 above)
Comment (wence-) [reply python/cudf/cudf/_lib/copying.pyx:157]: https://github.com/rapidsai/cudf/issues/12135

**Pull request merged**